cmd/istioctl/register.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strconv"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"istio.io/pilot/platform/kube"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	registerCmd = &cobra.Command{
+		Use:   "register <svcname> <ip> <ports...>",
+		Short: "Registers a service instance (VM)",
+		Args:  cobra.MinimumNArgs(3),
+		RunE: func(c *cobra.Command, args []string) error {
+			svcName := args[0]
+			ip := args[1]
+			portsListStr := args[2:]
+			portsList := make([]int32, len(portsListStr))
+			for i := range portsListStr {
+				p, err := strconv.Atoi(portsListStr[i])
+				if err != nil {
+					return err
+				}
+				portsList[i] = int32(p)
+			}
+			glog.Infof("Registering for service '%s' ip '%s', ports list %v",
+				svcName, ip, portsList)
+			return registerSvc(svcName, ip, portsList)
+		},
+	}
+)
```

Comment:
Port names are essential if you have more than one port for the vm. I think it's a k8s requirement. Also, unless you mark a port as http, pilot will not generate a http route config for that port. It will always be tcp which loses a ton of benefits.